### PR TITLE
Add a compiler intrinsic to back `bigint_helper_methods`

### PR DESCRIPTION
**Pull request by scottmcm**
cc https://github.com/rust-lang/rust/issues/85532

This adds a new `carrying_mul_add` intrinsic, to implement `wide_mul` and `carrying_mul`.

It has fallback MIR for all types -- including `u128`, which isn't currently supported on nightly -- so that it'll continue to work on all backends, including CTFE.

Then it's overridden in `cg_llvm` to use wider intermediate types, including `i256` for `u128::carrying_mul`.

### CODE REVIEW

File: library/core/src/intrinsics/fallback.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+#[rustc_const_unstable(feature = "core_intrinsics_fallbacks", issue = "none")]
+const fn wide_mul_u128(a: u128, b: u128) -> (u128, u128) {
+    const fn to_low_high(x: u128) -> [u64; 2] {
+        [x as u64, (x >> 64) as u64]
+    }
+    const fn from_low_high(x: [u64; 2]) -> u128 {
+        (x[0] as u128) | ((x[1] as u128) << 64)
+    }
+    #[rustc_const_unstable(feature = "core_intrinsics_fallbacks", issue = "none")]
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/intrinsics/fallback.rs:42]: If it compiles, it should be fine. ;)

But oh my this is hacky.^^

[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/intrinsics/fallback.rs:42]: @compiler-errors is currently working on some very unstable const trait support. This can't be used in stable coe any time soon, but if that's an acceptable constraint it may be worth playing around with that?

**Comment by clarfonthey** [comment]
Will take a look at this later, but I like that you managed to make fallback intrinsics work here when I couldn't quite get them to work for #132195.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 2]
```diff
@@ -340,6 +340,33 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
                     self.const_i32(cache_type),
                 ])
             }
+            sym::carrying_mul_add => {
+                let uty = fn_args.type_at(0);
+                let (size, signed) = uty.int_size_and_signed(self.tcx);
+                assert!(!signed);
```
Comment (clarfonthey) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:346]: Is there a particular reason why you exclude signedness here? I don't see anything particular here that depends on the integers being unsigned, since you could easily change `zext` into `intcast` to make this work like I did in my implementation: https://github.com/rust-lang/rust/pull/131566/files#diff-a762eadbc5da5c5fce9dff3c4a11e4ad720e6580767cc097d49e64264858aa61R463-R494

(not sure if this affects codegen, though, since I never tested that)

File: library/core/src/intrinsics/fallback.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (oli-obk) [reply library/core/src/intrinsics/fallback.rs:42]: I think it's already turned on in libcore and should be available without bootstrap shenanigans

**Comment by clarfonthey** [comment]
Also just to clarify my opinion on where this stands relative to #132195: I think that this PR just adding the intrinsic is okay, although I would like the intrinsic to also handle the signed case, even if it's only used at the moment for the unsigned case. That way, my PR can just update the library methods to use the intrinsic, without having to worry about the slow implementation.

Although the signed version of the *library* methods will likely return an unsigned word always for the least-significant / high-order part, it vastly simplifies the intrinsic to just have the same type for the arguments and return value, so, I would prefer that we do this specifically for the intrinsic. Since it doesn't actually affect the returned bits, I think that it's okay to just cast half of it unsigned on the library side. You can see how I did this [here](https://github.com/rust-lang/rust/pull/131566/files#diff-a762eadbc5da5c5fce9dff3c4a11e4ad720e6580767cc097d49e64264858aa61R463-R494) in a previous (closed) PR.

Otherwise, LGTM. Also, while the codegen test should suffice, if you want to copy over the library tests I added in my PR to verify that the intrinsic works correctly, feel free. I specifically added a bunch of edge cases with `MAX` and `MIN` as operands to ensure that things work correctly.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs (hunk 2 above)
Comment (clarfonthey) [reply compiler/rustc_codegen_llvm/src/intrinsic.rs:346]: Ah! Fair; I did manage to make it work in my purely library solution, but haven't since compared the two. I mostly just special-cased the signed versions.

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 3]
```diff
@@ -340,6 +340,42 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
...
+            sym::carrying_mul_add => {
+                let (size, signed) = fn_args.type_at(0).int_size_and_signed(self.tcx);
+
+                let wide_llty = self.type_ix(size.bits() * 2);
+                let [a, b, c, d] = args.as_array().unwrap().map(|a| {
+                    if signed {
+                        self.sext(a.immediate(), wide_llty)
+                    } else {
+                        self.zext(a.immediate(), wide_llty)
+                    }
```
Comment (clarfonthey) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:352]: Small nit, but isn't this the same as `self.intcast(a.immediate(), wide_llty, signed)` ?

File: compiler/rustc_codegen_llvm/src/intrinsic.rs [hunk 4]
```diff
@@ -340,6 +340,42 @@ impl<'ll, 'tcx> IntrinsicCallBuilderMethods<'tcx> for Builder<'_, 'll, 'tcx> {
...
+                let narrow_llty = self.type_ix(size.bits());
+                let low = self.trunc(wide, narrow_llty);
+                let bits_const = self.const_uint(wide_llty, size.bits());
+                // No need for ashr when signed; LLVM changes it to lshr anyway.
+                let high = self.lshr(wide, bits_const);
+                // FIXME: could be `trunc nuw`, even for signed.
+                let high = self.trunc(high, narrow_llty);
+
+                let pair_llty = self.type_struct(&[narrow_llty, narrow_llty], false);
+                let pair = self.const_poison(pair_llty);
```
Comment (clarfonthey) [inline compiler/rustc_codegen_llvm/src/intrinsic.rs:374]: Just comparing to my version, this doesn't *really* matter that much, but `const_undef` feels slightly more accurate here just semantically. Not actually sure if this affects codegen though.

**Comment by ibraheemdev** [comment]
This is outside my territory so I'm going to pass this along. r? libs

**Comment by Amanieu** [comment]
Is it intentional that this intrinsic is only ever called with `addend == 0 `? Are there plans to use the addend in other places in the future? If not then the intrinsic could probably be simplified.

Otherwise LGTM

**Comment by clarfonthey** [comment]
Yes; once this is merged I'll be rebasing #132195 on top of it which does add methods that use this to its full extent. I think that scottmcm just wanted to lower the scope of this PR to just the intrinsic, and not any library changes.
